crates/ide-db/src/search.rs

```diff
@@ -442,6 +447,400 @@ impl<'a> FindUsages<'a> {
+                                    if !contains_self {
+                                        continue;
+                                    }
+                                    if seen.insert(InFileWrapper::new(
+                                        file_id,
+                                        name.syntax().text_range(),
+                                    )) {
+                                        if let Some(def) = is_alias(&type_alias) {
+                                            insert_type_alias(
+                                                sema.db,
+                                                &mut to_process,
+                                                name.text().as_str(),
+                                                def.into(),
+                                            );
+                                        }
+                                    }
+                                }
+                                Some(())
+                            })();
+                        }
+                    }
+                }
+
+                completed.insert(current_to_process);
+
+                total_files_searched += current_to_process_search_scope.entries.len();
+                // FIXME: Maybe this needs to be relative to the project size, or at least to the initial search scope?
+                if total_files_searched > 20_000 && completed.len() > 100 {
+                    // This case is extremely unlikely (even searching for `Vec::new()` on rust-analyzer does not enter
+                    // here - it searches less than 10,000 files, and it does so in five seconds), but if we get here,
+                    // we at a risk of entering an almost-infinite loop of growing the aliases list. So just stop and
+                    // let normal search handle this case.
+                    tracing::info!(aliases_count = %completed.len(), "too much aliases; leaving fast path");
+                    return None;
+                }
+            }
+
+            // Impls can contain each other, so we need to deduplicate their ranges.
+            is_possibly_self.sort_unstable_by_key(|position| {
+                (position.file_id, position.range.start(), Reverse(position.range.end()))
+            });
+            is_possibly_self.dedup_by(|pos2, pos1| {
+                pos1.file_id == pos2.file_id
+                    && pos1.range.start() <= pos2.range.start()
+                    && pos1.range.end() >= pos2.range.end()
+            });
+
+            tracing::info!(aliases_count = %completed.len(), "aliases search completed");
+
+            Some((completed, is_possibly_self))
+        }
+
+        fn search(
+            this: &FindUsages<'_>,
+            finder: &Finder<'_>,
+            name: &str,
+            files: impl Iterator<Item = (Arc<str>, EditionedFileId, TextRange)>,
+            mut container_predicate: impl FnMut(
+                &SyntaxNode,
```

Comment:
We can keep it as `ast::Path`, but this will give us nothing. We will still need to use the raw tokens API, because if we search for plain path only, we will miss cases like `type Itself<T> = T; Itself<Self>`. We need to check if the identifier appears *anywhere* in the path.

Reply:
Ah okay, I misunderstood the idea behind this